### PR TITLE
refactor(cuda): introduce scratch for amortized PBS

### PR DESCRIPTION
**Pull request by agnesLeroy**
### Checklist 

(Use '[x]' to check the checkboxes, or submit the PR and then click the checkboxes)

* [ ] Tests for the changes have been added (for bug fixes / features)
* [ ] Docs have been added / updated (for bug fixes / features)
* [ ] The PR description links to the related issue (to link an issue, use '#XXX'.)
* [ ] The tests on AWS have been launched and are successful (comment with @slab-ci cpu_test and/or @slab-ci gpu_test to trigger the tests)
* [ ] The draft release description has been updated
* [ ] Check for breaking changes (including serialization changes) and add them to commit message following the conventional commit [specification][conventional-breaking]

[conventional-breaking]: https://www.conventionalcommits.org/en/v1.0.0/#commit-message-with-description-and-breaking-change-footer

### CODE REVIEW

**Comment by agnesLeroy** [comment]
@slab-ci gpu_test

**Comment by agnesLeroy** [comment]
@slab-ci gpu_test

**Comment by agnesLeroy** [comment]
@slab-ci gpu_test

File: concrete-cuda/cuda/src/bootstrap_amortized.cu [hunk 1]
```diff
@@ -1,15 +1,103 @@
...
+void scratch_cuda_bootstrap_amortized_32(void *v_stream, uint32_t gpu_index,
+                                         int8_t **pbs_buffer,
+                                         uint32_t glwe_dimension,
+                                         uint32_t polynomial_size,
+                                         uint32_t input_lwe_ciphertext_count,
+                                         uint32_t max_shared_memory,
+                                         bool allocate_gpu_memory) {
+
+  switch (polynomial_size) {
+  case 512:
```
Comment (pdroalves) [inline concrete-cuda/cuda/src/bootstrap_amortized.cu:17]: We already support `256`. We need a new clause in this `switch`.

File: concrete-cuda/cuda/src/bootstrap_amortized.cu [hunk 2]
```diff
@@ -1,15 +1,103 @@
...
+void scratch_cuda_bootstrap_amortized_64(void *v_stream, uint32_t gpu_index,
+                                         int8_t **pbs_buffer,
+                                         uint32_t glwe_dimension,
+                                         uint32_t polynomial_size,
+                                         uint32_t input_lwe_ciphertext_count,
+                                         uint32_t max_shared_memory,
+                                         bool allocate_gpu_memory) {
+
+  switch (polynomial_size) {
+  case 512:
```
Comment (pdroalves) [inline concrete-cuda/cuda/src/bootstrap_amortized.cu:61]: Here too.

File: concrete-cuda/cuda/src/bootstrap_amortized.cuh [hunk 3]
```diff
@@ -213,36 +213,107 @@ __global__ void device_bootstrap_amortized(
                                      glwe_dimension);
 }
 
+template <typename Torus>
+__host__ __device__ int
+get_full_sm_bootstrap_amortized(uint32_t polynomial_size,
```
Comment (pdroalves) [inline concrete-cuda/cuda/src/bootstrap_amortized.cuh:218]: May we rename this function to `get_buffer_size_full_sm_bootstrap_amortized`?

File: concrete-cuda/cuda/src/bootstrap_amortized.cuh [hunk 4]
```diff
@@ -213,36 +213,107 @@ __global__ void device_bootstrap_amortized(
...
+             (glwe_dimension + 1) + // accumulator rotated
+         sizeof(double2) * polynomial_size / 2 *
+             (glwe_dimension + 1) + // accumulator fft
+         sizeof(double2) * polynomial_size / 2 *
+             (glwe_dimension + 1); // calculate buffer fft
+}
+
+template <typename Torus>
+__host__ __device__ int
+get_partial_sm_bootstrap_amortized(uint32_t polynomial_size,
```
Comment (pdroalves) [inline concrete-cuda/cuda/src/bootstrap_amortized.cuh:231]: May we rename this function to `get_buffer_size_partial_sm_bootstrap_amortized`?

**Review by pdroalves — CHANGES_REQUESTED** [review]
Everything looks good. I think we only need to add support to N=256 on the amortized PBS scratch function.

File: concrete-cuda/cuda/src/bootstrap_amortized.cuh (hunk 3 above)
Comment (agnesLeroy) [reply concrete-cuda/cuda/src/bootstrap_amortized.cuh:218]: Sure!

File: concrete-cuda/cuda/src/bootstrap_amortized.cuh (hunk 4 above)
Comment (agnesLeroy) [reply concrete-cuda/cuda/src/bootstrap_amortized.cuh:231]: Np

**Comment by agnesLeroy** [comment]
@slab-ci gpu_test

**Review by pdroalves — APPROVED** [review]

**Pull request merged**